File: app/src/main/java/com/komica/reader/ThreadDetailActivity.java
package com.komica.reader;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.komica.reader.adapter.PostAdapter;
import com.komica.reader.model.Post;
import com.komica.reader.model.Thread;
import com.komica.reader.service.KomicaService;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import androidx.appcompat.widget.Toolbar;

import com.komica.reader.viewmodel.ThreadDetailViewModel;
import com.komica.reader.viewmodel.ThreadDetailViewModelFactory;
import com.komica.reader.util.KLog;
import androidx.lifecycle.ViewModelProvider;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

public class ThreadDetailActivity extends AppCompatActivity {

    private TextView threadTitle;
    private RecyclerView recyclerView;
    private ProgressBar progressBar;
    private FloatingActionButton fabScrollDown;
    private FloatingActionButton fabShare;
    private FloatingActionButton fabReply;
    
    private View previewCard;
    private TextView previewTitle;
    private TextView previewContent;
    
    private PostAdapter adapter;
    private ThreadDetailViewModel viewModel;
    private Thread currentThread;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_thread_detail);

        if (getIntent().hasExtra("thread")) {
            currentThread = (Thread) getIntent().getSerializableExtra("thread");
        } else if (getIntent().hasExtra("thread_url")) {
            String url = getIntent().getStringExtra("thread_url");
            String title = getIntent().getStringExtra("thread_title");
            // Create a lightweight thread object
            currentThread = new Thread(
                String.valueOf(System.currentTimeMillis()), // temp id
                title != null ? title : "Loading...",
                "",
                0,
                url
            );
        }

        if (currentThread == null) {
            finish();
            return;
        }

        ThreadDetailViewModelFactory factory = new ThreadDetailViewModelFactory(getApplication(), currentThread);
        viewModel = new ViewModelProvider(this, factory).get(ThreadDetailViewModel.class);
        
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayShowTitleEnabled(false);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        threadTitle = findViewById(R.id.threadTitle);
        recyclerView = findViewById(R.id.recyclerView);
        progressBar = findViewById(R.id.progressBar);
        fabScrollDown = findViewById(R.id.fabScrollDown);
        fabShare = findViewById(R.id.fabShare);
        fabReply = findViewById(R.id.fabReply);
        
        previewCard = findViewById(R.id.previewCard);
        previewTitle = findViewById(R.id.previewTitle);
        previewContent = findViewById(R.id.previewContent);

        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        fabScrollDown.setOnClickListener(v -> scrollToBottom());
        fabShare.setOnClickListener(v -> shareThread());
        fabReply.setOnClickListener(v -> showReplyDialog());

        threadTitle.setText(currentThread.getTitle());
        
        observeViewModel();
    }

    private void observeViewModel() {
        viewModel.getThreadDetail().observe(this, thread -> {
            if (thread != null) {
                updateUI(thread);
            }
        });

        viewModel.getIsLoading().observe(this, isLoading -> {
            progressBar.setVisibility(isLoading ? View.VISIBLE : View.GONE);
        });

        viewModel.getErrorMessage().observe(this, error -> {
            if (error != null) {
                android.widget.Toast.makeText(this, error, android.widget.Toast.LENGTH_LONG).show();
            }
        });

        viewModel.getReplyStatus().observe(this, success -> {
            if (success != null) {
                if (success) {
                    android.widget.Toast.makeText(this, "回覆成功", android.widget.Toast.LENGTH_SHORT).show();
                    scrollToBottom();
                } else {
                    android.widget.Toast.makeText(this, "回覆失敗", android.widget.Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

    private void showReplyDialog() {
        android.app.AlertDialog.Builder builder = new android.app.AlertDialog.Builder(this);
        View view = getLayoutInflater().inflate(R.layout.dialog_reply, null);
        
        com.google.android.material.textfield.TextInputEditText editContent = view.findViewById(R.id.editContent);
        Button btnCancel = view.findViewById(R.id.btnCancel);
        Button btnSend = view.findViewById(R.id.btnSend);
        
        builder.setView(view);
        android.app.AlertDialog dialog = builder.create();
        
        btnCancel.setOnClickListener(v -> dialog.dismiss());
        
        btnSend.setOnClickListener(v -> {
            String content = editContent.getText() != null ? editContent.getText().toString() : "";
            if (content.trim().isEmpty()) {
                editContent.setError("內容不能為空");
                return;
            }
            
            viewModel.sendReply(content);
            dialog.dismiss();
        });
        
        dialog.show();
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (adapter != null) {
            adapter.notifyDataSetChanged();
        }
    }

    private void updateUI(Thread thread) {
        List<Post> posts = thread.getPosts();
        adapter = new PostAdapter(posts, new PostAdapter.OnQuoteInteractionListener() {
            @Override
            public void onImageClick(int imageIndex, List<String> imageUrls) {
                Intent intent = new Intent(ThreadDetailActivity.this, ImagePreviewActivity.class);
                intent.putStringArrayListExtra("imageUrls", new ArrayList<>(imageUrls));
                intent.putExtra("position", imageIndex);
                startActivity(intent);
            }

            @Override
            public void onImageLongClick(String imageUrl) {
                shareImage(imageUrl);
            }

            @Override
            public void onQuoteClick(int position) {
                recyclerView.smoothScrollToPosition(position);
            }

            @Override
            public void onQuoteLongClick(Post post) {
                if (post != null) {
                    previewTitle.setText("No. " + post.getNumber() + " " + (post.getAuthor() != null ? post.getAuthor() : ""));
                    previewContent.setText(post.getContent());
                    previewCard.setVisibility(View.VISIBLE);
                }
            }

            @Override
            public void onQuoteReleased() {
                previewCard.setVisibility(View.GONE);
            }
        });
        
        recyclerView.setAdapter(adapter);
    }

    private void scrollToBottom() {
        if (adapter != null) {
            int itemCount = adapter.getItemCount();
            if (itemCount > 0) {
                recyclerView.smoothScrollToPosition(itemCount - 1);
            }
        }
    }

    private void shareThread() {
        if (currentThread != null) {
            Intent shareIntent = new Intent(Intent.ACTION_SEND);
            shareIntent.setType("text/plain");
            
            String shareText = currentThread.getTitle() + "\n\n" + KomicaService.resolveUrl(currentThread.getUrl(), "");
            shareIntent.putExtra(Intent.EXTRA_TEXT, shareText);
            
            Intent chooserIntent = Intent.createChooser(shareIntent, "分享討論串");
            startActivity(chooserIntent);
        }
    }

    private void shareImage(String imageUrl) {
        com.komica.reader.repository.KomicaRepository.getInstance(this).execute(() -> {
            try {
                java.io.File file = com.bumptech.glide.Glide.with(this)
                    .asFile()
                    .load(imageUrl)
                    .submit()
                    .get();
                    
                android.net.Uri uri = androidx.core.content.FileProvider.getUriForFile(this, 
                    getApplicationContext().getPackageName() + ".fileprovider", file);
                    
                runOnUiThread(() -> {
                    Intent shareIntent = new Intent(Intent.ACTION_SEND);
                    shareIntent.setType("image/*");
                    shareIntent.putExtra(Intent.EXTRA_STREAM, uri);
                    shareIntent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
                    startActivity(Intent.createChooser(shareIntent, "分享圖片"));
                });
            } catch (Exception e) {
                e.printStackTrace();
                runOnUiThread(() -> android.widget.Toast.makeText(this, "分享圖片失敗", android.widget.Toast.LENGTH_SHORT).show());
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(android.view.Menu menu) {
        getMenuInflater().inflate(R.menu.thread_detail_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(android.view.MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        } else if (item.getItemId() == R.id.action_gallery) {
            if (currentThread != null) {
                Intent intent = new Intent(this, GalleryActivity.class);
                intent.putExtra("thread_url", currentThread.getUrl());
                intent.putExtra("thread_title", currentThread.getTitle());
                startActivity(intent);
            }
            return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/komica/reader/ThreadDetailActivity.kt
package com.komica.reader

import android.content.Intent
import android.app.Activity
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.activity.viewModels
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.komica.reader.adapter.PostAdapter
import com.komica.reader.data.AppSettings
import com.komica.reader.data.HistoryStore
import com.komica.reader.data.ImageStore
import com.komica.reader.databinding.ActivityThreadDetailBinding
import com.komica.reader.model.ThreadDetail
import com.komica.reader.model.KomicaThread
import com.komica.reader.model.UiState
import com.komica.reader.util.GetSerializableCompat
import com.komica.reader.util.WindowInsetsUtil
import com.komica.reader.viewmodel.ThreadDetailViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class ThreadDetailActivity : AppCompatActivity() {
    private lateinit var binding: ActivityThreadDetailBinding
    private val viewModel: ThreadDetailViewModel by viewModels()
    private lateinit var adapter: PostAdapter
    private var thread: KomicaThread? = null
    private var detail: ThreadDetail? = null
    private var hasLoadedOnce = false
    private val replyLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode == Activity.RESULT_OK) {
            thread?.let { viewModel.LoadThread(it.url) }
            ScrollToBottom()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        AppSettings(this).ApplyTheme()
        super.onCreate(savedInstanceState)
        binding = ActivityThreadDetailBinding.inflate(layoutInflater)
        setContentView(binding.root)
        WindowInsetsUtil.ApplyBrandStatusBar(window)
        WindowInsetsUtil.ApplyToolbarInsets(binding.toolbar)

        thread = intent.GetSerializableCompat(ExtraThread)
        if (thread == null) {
            finish()
            return
        }

        adapter = PostAdapter(
            postFontSizeSp = AppSettings(this).postFontSizeSp,
            onQuoteClick = { position -> ScrollToPost(position) },
            onQuotePreviewShow = { post, anchor, x, y -> ShowPostPreview(post.number, post.author, post.content, anchor, x, y) },
            onQuotePreviewHide = { HidePostPreview() },
            onImageClick = { position, urls ->
                startActivity(Intent(this, ImagePreviewActivity::class.java).apply {
                    putStringArrayListExtra(ImagePreviewActivity.ExtraImageUrls, ArrayList(urls))
                    putExtra(ImagePreviewActivity.ExtraPosition, position)
                })
            },
            onImageLongClick = { imageUrl ->
                startActivity(Intent(this, ImagePreviewActivity::class.java).apply {
                    putStringArrayListExtra(ImagePreviewActivity.ExtraImageUrls, arrayListOf(imageUrl))
                    putExtra(ImagePreviewActivity.ExtraPosition, 0)
                })
            }
        )

        binding.toolbar.title = thread?.title ?: "討論串"
        binding.toolbar.inflateMenu(R.menu.thread_detail_toolbar_menu)
        binding.toolbar.setOnMenuItemClickListener {
            when (it.itemId) {
                R.id.action_image_wall -> {
                    OpenImageWall()
                    true
                }
                R.id.action_download_all -> {
                    DownloadAllImages()
                    true
                }
                else -> false
            }
        }
        binding.fabReply.setOnClickListener { OpenReplyPage() }
        binding.fabShare.setOnClickListener { ShareThread() }
        binding.fabScrollDown.setOnClickListener { ScrollToBottom() }
        binding.postRecyclerView.layoutManager = LinearLayoutManager(this)
        binding.postRecyclerView.adapter = adapter
        binding.swipeRefreshLayout.setColorSchemeResources(R.color.kr_primary)
        binding.swipeRefreshLayout.setOnRefreshListener {
            thread?.let { viewModel.LoadThread(it.url) }
        }

        ObserveState()
    }

    override fun onResume() {
        super.onResume()
        thread?.let {
            viewModel.LoadThread(it.url)
            hasLoadedOnce = true
        }
    }

    private fun ObserveState() {
        viewModel.state.observe(this) { state ->
            binding.progressBar.visibility = if (state is UiState.Loading) View.VISIBLE else View.GONE
            binding.swipeRefreshLayout.isRefreshing = state is UiState.Loading && hasLoadedOnce
            binding.messageText.visibility = if (state is UiState.Empty || state is UiState.Error) View.VISIBLE else View.GONE
            binding.postRecyclerView.visibility = if (state is UiState.Content) View.VISIBLE else View.GONE

            when (state) {
                is UiState.Content -> {
                    detail = state.data
                    binding.toolbar.title = state.data.title
                    adapter.SubmitPosts(state.data.posts)
                    thread?.let { HistoryStore(this).Add(it.copy(title = state.data.title)) }
                }
                is UiState.Empty -> binding.messageText.text = state.message
                is UiState.Error -> {
                    binding.messageText.text = "${state.message}\n\n點擊重試"
                    binding.messageText.setOnClickListener { thread?.let { viewModel.LoadThread(it.url) } }
                }
                UiState.Loading -> Unit
            }
        }
    }

    private fun ScrollToPost(position: Int) {
        binding.postRecyclerView.scrollToPosition(position)
        binding.postRecyclerView.post {
            binding.postRecyclerView.findViewHolderForAdapterPosition(position)?.itemView?.let { view ->
                val original = view.background
                view.setBackgroundColor(ContextCompat.getColor(this, R.color.kr_surface_alt))
                view.postDelayed({ view.background = original }, 900L)
            }
        }
    }

    private fun ScrollToBottom() {
        val count = adapter.itemCount
        if (count > 0) binding.postRecyclerView.smoothScrollToPosition(count - 1)
    }

    private fun ShowPostPreview(number: Int, author: String, content: String, anchor: View, touchX: Float, touchY: Float) {
        binding.previewTitle.text = "No.$number $author"
        binding.previewContent.text = content.ifBlank { "(無文字內容)" }
        binding.previewCard.visibility = View.VISIBLE
        binding.previewCard.post {
            val rootLocation = IntArray(2)
            val anchorLocation = IntArray(2)
            binding.root.getLocationOnScreen(rootLocation)
            anchor.getLocationOnScreen(anchorLocation)

            val pointerX = anchorLocation[0] - rootLocation[0] + touchX
            val pointerY = anchorLocation[1] - rootLocation[1] + touchY
            val margin = resources.displayMetrics.density * 12f
            val previewWidth = binding.previewCard.width.toFloat()
            val previewHeight = binding.previewCard.height.toFloat()
            val maxX = binding.root.width - previewWidth - margin
            val targetX = (pointerX - previewWidth / 2f).coerceIn(margin, maxX.coerceAtLeast(margin))
            val targetY = (pointerY - previewHeight - margin).coerceAtLeast(margin)

            binding.previewCard.x = targetX
            binding.previewCard.y = targetY
        }
    }

    private fun HidePostPreview() {
        binding.previewCard.visibility = View.GONE
    }

    private fun CurrentImageUrls(): List<String> {
        return detail?.posts.orEmpty().mapNotNull { it.imageUrl.ifBlank { it.thumbnailUrl }.takeIf(String::isNotBlank) }
    }

    private fun OpenImageWall() {
        val urls = CurrentImageUrls()
        if (urls.isEmpty()) {
            Toast.makeText(this, "此討論串沒有圖片", Toast.LENGTH_SHORT).show()
            return
        }
        startActivity(Intent(this, ImageGridActivity::class.java).apply {
            putStringArrayListExtra(ImagePreviewActivity.ExtraImageUrls, ArrayList(urls))
        })
    }

    private fun DownloadAllImages() {
        val urls = CurrentImageUrls()
        if (urls.isEmpty()) {
            Toast.makeText(this, "此討論串沒有圖片", Toast.LENGTH_SHORT).show()
            return
        }
        lifecycleScope.launch {
            Toast.makeText(this@ThreadDetailActivity, "開始下載 ${urls.size} 張圖片", Toast.LENGTH_SHORT).show()
            val successCount = withContext(Dispatchers.IO) {
                urls.count { url -> runCatching { ImageStore.DownloadImage(this@ThreadDetailActivity, url) }.getOrDefault(false) }
            }
            Toast.makeText(this@ThreadDetailActivity, "已下載 $successCount / ${urls.size} 張圖片", Toast.LENGTH_LONG).show()
        }
    }

    private fun ShareThread() {
        val currentThread = thread ?: return
        val title = detail?.title ?: currentThread.title
        val intent = Intent(Intent.ACTION_SEND).apply {
            type = "text/plain"
            putExtra(Intent.EXTRA_SUBJECT, title)
            putExtra(Intent.EXTRA_TEXT, "$title\n${currentThread.url}")
        }
        startActivity(Intent.createChooser(intent, "分享討論串"))
    }

    private fun OpenReplyPage() {
        val currentThread = thread ?: return
        replyLauncher.launch(Intent(this, ReplyActivity::class.java).apply {
            putExtra(ReplyActivity.ExtraUrl, currentThread.url)
            putExtra(ReplyActivity.ExtraTitle, "回覆：${detail?.title ?: currentThread.title}")
        })
    }

    companion object {
        const val ExtraThread = "extra_thread"
    }
}
